src/main/java/org/traccar/handler/FilterHandler.java

```diff
@@ -167,19 +165,56 @@ private boolean filter(Position position) {
         if (filterApproximate(position)) {
             filterType.append("Approximate ");
         }
-        if (filterStatic(position) && !skipLimit(position, last) && !skipAttributes(position)) {
-            filterType.append("Static ");
-        }
-        if (filterDistance(position, last) && !skipLimit(position, last) && !skipAttributes(position)) {
-            filterType.append("Distance ");
-        }
-        if (filterMaxSpeed(position, last)) {
-            filterType.append("MaxSpeed ");
-        }
-        if (filterMinPeriod(position, last)) {
-            filterType.append("MinPeriod ");
-        }
+        if (filterRelative) {
+            try {
+                if (filterStatic(position) && !skipAttributes(position)) {
+                    filterType.append("Static ");
+                }
 
+                Date newfixTime = position.getFixTime();
+                Position duplicate = Context.getDataManager().getPositionByTime(position.getDeviceId(), newfixTime);
+                if (filterDuplicate(position, duplicate) && !skipLimit(position, duplicate)
+                        && !skipAttributes(position)) {
+                    filterType.append("Duplicate ");
+                }
+
+                Position previous = Context.getDataManager().getPrevPosition(position.getDeviceId(), newfixTime);
+                Position next = Context.getDataManager().getNextPosition(position.getDeviceId(), newfixTime);
```

Comment:
Can you please provide details on why we need "next" position.

Reply:
The backend has configured MinPeriod filter set to 2 seconds. Imagine that these  locations were received by the server:
* Location A: `fixTime = 18:18:01`
* Location C: `fixTime = 18:18:05`

Then after 10 seconds, server receives location B with `fixTime = 18:18:04`. As it turns out, it was not delivered immediately because device was offline for 3 seconds, because it was switching between networks.

Location B is far enough from location A, but it's too close to location C time-wise. Hence, we need to take into account _next_ position as well.

Reply:
It makes sense in theory, but I'm not sure if it makes sense to actually implement it. You basically make 3 expensive SQL requests for every location report. It will cripple pretty much any server that runs any significant number of devices.

Reply:
I agree that with thousands of tracking devices this will have deteriorating impact. But for private use it will not be a problem. No doubt, that we should make this option disabled by default.

What do you say, if I introduce this optimization: let's load _next_ and _prev_ positions only in case any of these filters are turned active:
- filterDistance
- filterMaxSpeed
- filterMinPeriod

So if we only duplicate filter is active, than we have only 1 request instead of 3.

Reply:
I'm mostly worried about value of this feature vs cost of maintaining it going forward.

Reply:
I thought [you were Okay](https://github.com/traccar/traccar/issues/4735#issuecomment-908474813) with me adding this logic turned OFF by default.

Reply:
Yeah, the discussion was about loading one last position, if I understood it correctly.

Reply:
I can rewrite the code so that it uses only 1 sql query:

`SELECT * FROM tc_positions WHERE deviceId = :deviceId AND NOT fixTime <= :time ORDER BY fixTime DESC  LIMIT 1`

Would it be satisfactory?

Reply:
Yes, that makes sense to me. In that case I think the logic would basically stay the same and you just change the way you get "last" position.

Reply:
Why do you use `last` to calculate speed, when you may take speed value from Position (if available)?

Reply:
It it used for filtering big jumps, like when device suddenly moves to a 0,0 location. Speed reported by device is not useful in those cases.

Reply:
Can we call `skipLimit` and `skipAttributes` only once in handleFilter()?

```
if (skipLimit(position, previous) || skipAttributes(position)) {
    return true;
}
```

Reply:
We can also make lazy loading of last / previous position. What do you say? Only in the case when at least one of the corresponding filters is activated.

Reply:
Makes sense to me.

Reply:
Check this out:
![image](https://user-images.githubusercontent.com/100644/132959567-7c21246b-2f3b-4471-800d-1a0e285999e6.png)

line 149: moved skipAttr to the top
line 176: lazy instantiation of _previous_

Reply:
Looks fine, but please update PR so I can review it more thoroughly.

Reply:
OK, I have pushed a new commit.